src/lseq.rs

```diff
@@ -36,95 +36,103 @@
+}
+
+impl<A: Ord> PartialOrd for Index<A> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(&other))
+    }
+}
+
+impl<A: Ord> Ord for Index<A> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        (&self.id, &self.dot.actor, &self.dot.counter).cmp(&(
+            &other.id,
+            &other.dot.actor,
+            &other.dot.counter,
+        ))
+    }
 }
 
 /// As described in the module documentation:
 ///
 /// An LSEQ tree is a CRDT for storing sequences of data (Strings, ordered lists).
 /// It provides an efficient view of the stored sequence, with fast index, insertion and deletion
 /// operations.
-#[derive(Clone, Serialize, Deserialize, PartialEq, Eq, Hash, PartialOrd)]
+#[derive(Clone, Serialize, Deserialize, PartialEq, Eq, Hash)]
 pub struct LSeq<T, A: Ord> {
-    seq: Vec<Entry<T, A>>,
-    gen: IdentGen<A>,
+    seq: BTreeMap<Index<A>, T>,
     clock: VClock<A>,
 }
 
 /// Operations that can be performed on an LSeq tree
-#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Hash, PartialOrd)]
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
 pub enum Op<T, A> {
     /// Insert an element
     Insert {
-        /// Identifier to insert at
-        id: Identifier<A>,
-        /// clock of site that issued insertion
-        dot: Dot<A>,
+        /// The Index to insert at
+        index: Index<A>,
         /// Element to insert
         val: T,
     },
     /// Delete an element
     Delete {
-        /// The original clock information of the insertion we're removing
-        remote: Dot<A>,
-        /// Identifier to remove
-        id: Identifier<A>,
+        /// The Index of the insertion we're removing
+        index: Index<A>,
         /// id of site that issued delete
         dot: Dot<A>,
     },
 }
 
 impl<T, A> Op<T, A> {
-    /// Return the Dot originating the operation
-    pub fn dot(&self) -> &Dot<A> {
+    /// Returns the Index this operation is concerning.
+    pub fn index(&self) -> &Index<A> {
         match self {
-            Op::Insert { dot, .. } | Op::Delete { dot, .. } => dot,
+            Op::Insert { index, .. } | Op::Delete { index, .. } => index,
         }
     }
 
-    /// Return the Identifier contained in the operation
-    pub fn id(&self) -> &Identifier<A> {
+    /// Return the Dot originating the operation.
+    pub fn dot(&self) -> &Dot<A> {
         match self {
-            Op::Insert { id, .. } | Op::Delete { id, .. } => id,
+            Op::Insert { index, .. } => &index.dot,
+            Op::Delete { dot, .. } => dot,
         }
     }
 }
 
-impl<T: Clone, A: Ord + Clone> LSeq<T, A> {
-    /// Create an empty LSEQ
-    pub fn new(id: A) -> Self {
-        LSeq {
-            seq: Vec::new(),
-            gen: IdentGen::new(id),
-            clock: VClock::new(),
+impl<T, A: Ord> Default for LSeq<T, A> {
+    fn default() -> Self {
+        Self {
+            seq: Default::default(),
+            clock: Default::default(),
```

Comment:
What's your opinion on not using `Default::default() but the specific type's `default()`? just curious as I thought it'd be better and more strict to use the specific type's one but...not sure...I guess it's not really a big deal in reality

Reply:
I get the argument for explicitness, but I feel at some point you need to trust the abstractions. The types are defined explicitly at the struct definition, here all I want to say is I want the default value of these fields, no matter what the type changes to in the future.